src/core/libraries/kernel/sync/semaphore.cpp

```diff
@@ -0,0 +1,113 @@
+// SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "semaphore.h"
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <semaphore>
+#endif
+
+namespace Libraries::Kernel {
+
+BinarySemaphore::BinarySemaphore() {
+#ifdef _WIN64
+    sem = CreateSemaphore(nullptr, 1, 1, nullptr);
+    ASSERT(sem);
+#else
+    sem = std::binary_semaphore(0);
+#endif
+}
+
+BinarySemaphore::~BinarySemaphore() {
+#ifdef _WIN64
+    CloseHandle(sem);
+#endif
+}
+
+void BinarySemaphore::release() {
+#ifdef _WIN64
+    ReleaseSemaphore(sem, 1, nullptr);
+#else
+    sem.release();
+#endif
+}
+
+void BinarySemaphore::acquire() {
+#ifdef _WIN64
+    for (;;) {
+        u64 res = WaitForSingleObjectEx(sem, INFINITE, true);
+        if (res == WAIT_OBJECT_0) {
+            return;
+        }
+    }
+#else
+    sem.acquire();
+#endif
+}
+
+bool BinarySemaphore::try_acquire() {
+#ifdef _WIN64
+    return WaitForSingleObjectEx(sem, 0, true) == WAIT_OBJECT_0;
+#else
+    return sem.try_acquire();
+#endif
+}
+
+bool BinarySemaphore::try_acquire_for(std::chrono::microseconds timeout) {
+#ifdef _WIN64
+    u64 timeout_ms = std::chrono::duration_cast<std::chrono::milliseconds>(timeout).count();
+    if (timeout_ms == 0) {
+        return try_acquire();
+    }
+
+    return WaitForSingleObjectEx(sem, timeout_ms, true) == WAIT_OBJECT_0;
+#else
+    return sem.try_acquire_for(timeout);
+#endif
+}
+
+CountingSemaphore::CountingSemaphore(s32 initialCount, s32 maxCount) {

```

Comment:
This seems redundant, why can't we just have a single semaphore implementation and use initial/max count = 1 for binary semaphore?

Reply:
They use `binary_semaphore` and `counting_semaphore` from the STL on other platforms and their implementation might be different (it is on MSVC).

Reply:
They should still behave equivalent though, on other platforms `binary_semaphore` is just a typedef for `counting_semaphore<1>`

Reply:
At least on msvc stl binary_semaphore isnt a typedef but has actual special implementation iirc